metrics/engine/ingester.go

```diff
@@ -92,6 +97,74 @@ func (oi *outputIngester) flushMetrics() {
 				oi.metricsEngine.markObserved(sm.Metric)
 				sm.Metric.Sink.Add(sample)
 			}
+
+			oi.cardinality.Add(sample.TimeSeries)
 		}
 	}
+
+	if hit, limit := oi.cardinality.LimitHit(); hit {
+		// TODO: suggest using the Metadata API as an alternative, once it's
+		// available (e.g. move high-cardinality tags as Metadata)
+		// https://github.com/grafana/k6/issues/2766
+
+		oi.logger.Warnf("The test has generated %d time series, which is higher than the suggested limit (%d), "+
+			"and could cause high memory usage. "+
+			"Consider not using high-cardinality values like IDs as tags "+
+			"or if you need them in the URL then use the URL-grouping feature. "+
+			"See https://k6.io/docs/using-k6/tags-and-groups for details.", oi.cardinality.Count(), limit)
+	}
+}
+
+type cardinalityControl struct {
+	seen            map[metrics.TimeSeries]struct{}
+	timeSeriesLimit int
+	addsCount       int
+}
+
+func newCardinalityControl() *cardinalityControl {
+	return &cardinalityControl{
+		timeSeriesLimit: timeSeriesFirstLimit,
+		addsCount:       0,
+		seen:            make(map[metrics.TimeSeries]struct{}),
+	}
+}
+
+// Add adds the passed time series to the list of seen items.
+func (cc *cardinalityControl) Add(ts metrics.TimeSeries) {
+	cc.addsCount++
+	cc.seen[ts] = struct{}{}
```

Comment:
hmm, shouldn't this be something like this:
```suggestion
	if _, ok := cc.seen[ts]; !ok {
		cc.addsCount++
		cc.seen[ts] = struct{}{}
	}
```

a test that we don't count the same time series twice would be nice too